pubsub/gcp/pubsub/pubsub.go

```diff
@@ -354,9 +416,21 @@ func (g *GCPPubSub) getTopic(topic string) *gcppubsub.Topic {
 }
 
 func (g *GCPPubSub) ensureSubscription(parentCtx context.Context, subscription string, topic string) error {
-	err := g.ensureTopic(parentCtx, topic)
-	if err != nil {
-		return err
+	g.lock.RLock()
+	_, exists := g.topicCache[topic]
+	_, DExists := g.topicCache[g.metadata.DeadLetterTopic]
+	g.lock.RUnlock()
+	if !exists {
```

Comment:
This is almost good. The problem here is if `ensureSubscription` is called by 2 callers in parallel:

1. Caller A will get a read lock to check that the topic exists
2. Caller B does the same
3. Both see the topic does NOT exist
4. Both will create the topic
5. Both will save it in `topicCache

This is a bit wasteful.

The common pattern is:

1. Acquire a read lock and check if the topic exists
2. If the topic exists, all good
3. If it does NOT exist, acquire a write lock **before** calling `g.ensureTopic`. This means that there won't be > 1 call to `ensureTopic` at the same time
4. After getting the write lock, **before** calling `g.ensureTopic`, check again if the value exists in the cache. This is because by the time we acquire the lock, someone else may have created the topic in the cache.

You can find multiple examples of this pattern in the codebase!

Same feedback applies to the other instances :)

Reply:
Done ! Is it because this function is bit long so it might be prone to happening like that ? because or for every function we might endup doing like first ```RLock``` and then for updating we read it again ?